File: src/main/java/MaxProfitSolution.java
public class MaxProfitSolution {
    public int maxProfitSolution(int[] prices) {

        int currentProfit = 0;
        int currentMinPrice = Integer.MAX_VALUE;

        for (int price : prices) {
            if (price < currentMinPrice) {
                currentMinPrice = price;

            } else {
                int possibleProfit = price - currentMinPrice;
                if (possibleProfit > currentProfit) {
                    currentProfit = possibleProfit;
                }
            }
        }

        return currentProfit;
    }


    public static void main(String[] args) {

        MaxProfitSolution maxProfitSolution = new MaxProfitSolution();
        int[] array = {7, 1, 5, 3, 6, 4};
        int result = maxProfitSolution.maxProfitSolution(array);

        System.out.println("Result = " + result);

    }
}

File: src/main/kotlin/MaxProfitSolution.kt
class MaxProfitSolutionKotlin {
    fun maxProfit(prices: IntArray): Int {
        var lowestPrice = Int.MAX_VALUE
        var profit = 0
        prices.forEach { value ->
            if (value < lowestPrice) {
                lowestPrice = value
            } else {
                val currentProfit = value - lowestPrice
                if (currentProfit > profit) {
                    profit = currentProfit
                }
            }
        }
        return profit
    }
}

fun main() {
    val array = intArrayOf(7,1,5,3,6,4)
    val result = MaxProfitSolutionKotlin().maxProfit(array)

    println("Result = $result")
}